agent/checks/grpc.go

```diff
@@ -49,17 +50,25 @@ func NewGrpcHealthProbe(target string, timeout time.Duration, tlsConfig *tls.Con
 
 // Check if the target of this GrpcHealthProbe is healthy
 // If nil is returned, target is healthy, otherwise target is not healthy
-func (probe *GrpcHealthProbe) Check(target string) error {
+func (probe *GrpcHealthProbe) Check(target string, header map[string][]string) error {
 	serverAndService := strings.SplitN(target, "/", 2)
 	serverWithScheme := fmt.Sprintf("%s:///%s", resolver.GetDefaultScheme(), serverAndService[0])
 
-	ctx, cancel := context.WithTimeout(context.Background(), probe.timeout)
+	var headerMD map[string]string = make(map[string]string)
+	for k, v := range header {
+		headerMD[k] = strings.Join(v, ", ")
+	}
+
+	md := metadata.New(headerMD)
+
+	ctx, cancel := context.WithTimeout(metadata.NewOutgoingContext(context.Background(), md), probe.timeout)
```

Comment:
Can we conditionally run line 57 to line 64?

```go
if len(header) > 0 {
    // ...
    ctx, cancel := context.WithTimeout(metadata.NewOutgoingContext(context.Background(), md), probe.timeout)
}
```

Reply:
Hi, huikang! Thank you for your review! I think this should be changed because `ctx` is declared in that line and we need it in the next lines, for the `cancel` variable, instead, we need to call it in the `defer cancel() `, so I think this condition should go from line 57 to line 65 and declaring the `ctx` outside the scope of the `if len(header) > 0 ` block

Reply:
Thanks. By looking at the `grpc/metadata` package I think we can do the following

```go
ctx, cancel := context.WithTimeout(context.Background(), probe.timeout)

if len(header) > 0 {
    // line 57 - line 62
    // headerMD := map[string]string{}
    // we might need to change MD to kv...

    ctx = metadata.AppendToOutgoingContext(ctx, kv...) context.Context](https://pkg.go.dev/google.golang.org/grpc/metadata#AppendToOutgoingContext)
}
```
This way we can both conditionally add header to metadata, and reuse `ctx, and cancel`. WDTY